pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -170,18 +170,31 @@ def _su2su2_to_tensor_products(U):
     # C1 C2^dag = a1 a2* I
     C12 = math.dot(C1, math.conj(math.T(C2)))
 
-    if not math.allclose(a1 * math.conj(a2), C12[0, 0]):
-        a2 *= -1
+    if not math.is_abstract(C12):
+        if not math.allclose(a1 * math.conj(a2), C12[0, 0]):
+            a2 *= -1
+    else:
+        sign_is_correct = math.allclose(a1 * math.conj(a2), C12[0, 0])
+        sign = (-1) ** (sign_is_correct + 1)
+        a2 *= sign
 
     # Construct A
     A = math.stack([math.stack([a1, a2]), math.stack([-math.conj(a2), math.conj(a1)])])
 
     # Next, extract B. Can do from any of the C, just need to be careful in
     # case one of the elements of A is 0.
-    if not math.allclose(A[0, 0], 0.0, atol=1e-6):
-        B = C1 / math.cast_like(A[0, 0], 1j)
+    # We use B1 unless division by 0 has caused all elements to be inf.
+    if not math.is_abstract(A):
+        if not math.allclose(A[0, 0], 0.0, atol=1e-6):
+            B = C1 / math.cast_like(A[0, 0], 1j)
+        else:
+            B = C2 / math.cast_like(A[0, 1], 1j)
     else:
-        B = C2 / math.cast_like(A[0, 1], 1j)
+        B1 = C1 / math.cast_like(A[0, 0], 1j)
+        B2 = C2 / math.cast_like(A[0, 1], 1j)
+        B1_and_B2 = math.array([B1, B2], like="jax")
+        idx = math.allclose(A[0, 0], 0.0, atol=1e-6) + 0
```

Comment:
Minor: I wonder if "choosing" `B1` or `B2` could be done with an `if` statement of via `jax.lax.cond` - no real difference, but would perhaps improve readability.

Reply:
Good idea, thanks! Haven't seen that one before.